### ISSUE TITLE
Type CustomizableOptionInput supports only String! for value

### ISSUE THREAD

**Issue by TomashKhamlai** [magento/graphql-ce#345]
Test Executions:
[1](https://app.hiptest.com/projects/114163/test-runs/232368/folder-snapshots/1925571/scenario-snapshots/7470060/test-snapshots/9653675), [2](https://app.hiptest.com/projects/114163/test-runs/232368/folder-snapshots/1925571/scenario-snapshots/7470060/test-snapshots/9653676), [3](https://app.hiptest.com/projects/114163/test-runs/232368/folder-snapshots/1925571/scenario-snapshots/7470060/test-snapshots/9653677), [4](https://app.hiptest.com/projects/114163/test-runs/232368/folder-snapshots/1925571/scenario-snapshots/7470060/test-snapshots/9653678), [5](https://app.hiptest.com/projects/114163/test-runs/232368/folder-snapshots/1925571/scenario-snapshots/7470060/test-snapshots/9653679)
Scenario: [https://app.hiptest.com/projects/114163/test-plan/folders/801154/scenarios/2695714](https://app.hiptest.com/projects/114163/test-plan/folders/801154/scenarios/2695714)
Dataset: Checkbox, Multiple Select, Date, Date & Time, Time

### Preconditions (*)

1. Magento version 2.3.0

### Steps to reproduce (*)

1. Login to back-end as admin
2. Go to Catalog -> Products
3. Click "Add Product"
4. Fill in the fields
```
Required fields:
----------------------
Name: simple1
SKU: simple1
Price: 10 $

Not required fields:
----------------------------
Tax Group = Taxable Goods
Qty = 1000, In Stock
Weight = 1
```
5. Select category "cat1"
6. Click Save button
7. Expand "Customizable Options" tab
8. Click Add New Option
9. Add required Custom Option of **Checkbox** / **Multiple Select** / **Date** / **Date & Time** / **Time**  type
10. Select the input data from the datatable according to the input type
```
|=========================|==============================|
| INPUT TYPE              | Field                        |
| Option title            | Custom option field 1        |
| Required                | Yes                          |
| Price                   | 15                           |
| Price Type              | Fixed                        |
| SKU                     | sku_field                    |
| Max Characters          | 25                           |
|=========================|==============================|
| INPUT TYPE              | Area                         |
| Option title            | Custom option area 2         |
| Required                | Yes                          |
| Price                   | 50                           |
| Price Type              | Percent                      |
| SKU                     | sku_area                     |
| Max Characters          | 256                          |
|=========================|==============================|
| INPUT TYPE              | File                         |
| Option title            | Custom option file 3         |
| Required                | Yes                          |
| Price                   | 12                           |
| Price Type              | Fixed                        |
| SKU                     | sku_file                     |
| Allowed File Extensions | gif, bmp                     |
| Max Image Width         | 320                          |
| Max Image Heigth        | 240                          |
|=========================|==============================|
| INPUT TYPE              | Drop-down                    |
| Option title            | Custom option drop down 4    |
| Required                | Yes                          |
| Title                   | Title Drop-down 1            |
| Price                   | 2.56                         |
| Price Type              | Fixed                        |
| SKU                     | sku_drop_down_row_1          |
|=========================|==============================|
| INPUT TYPE              | Radio Button                 |
| Option title            | Custom option radio button 5 |
| Required                | Yes                          |
| Title                   | Title Radio Button 1         |
| Price                   | 2                            |
| Price Type              | Percent                      |
| SKU                     | sku_radio_row_1              |
|=========================|==============================|
| INPUT TYPE              | Checkbox                     |
| Option title            | Custom option checkbox 6     |
| Required                | Yes                          |
| Title                   | Title Checkbox 1             |
| Price                   | 10                           |
| Price Type              | Fixed                        |
| SKU                     | sku_checkbox_1               |
|=========================|==============================|
| INPUT TYPE              | Multiple Select              |
| Option title            | Custom option multiple 7     |
| Required                | Yes                          |
| Title                   | Title Multiple 1             |
| Price                   | 5                            |
| Price Type              | Percent                      |
| SKU                     | sku_multiple_1               |
|=========================|==============================|
| INPUT TYPE              | Date                         |
| Option title            | Custom option date 8         |
| Required                | Yes                          |
| Price                   | 5                            |
| Price Type              | Fixed                        |
| SKU                     | sku_date                     |
|=========================|==============================|
| INPUT TYPE              | Date & Time                  |
| Option title            | Custom option date & time 9  |
| Required                | Yes                          |
| Price                   | 13                           |
| Price Type              | Fixed                        |
| SKU                     | sku_date_time                |
|=========================|==============================|
| INPUT TYPE              | Time                         |
| Option title            | Custom option time 10        |
| Required                | Yes                          |
| Price                   | 14                           |
| Price Type              | Percent                      |
| SKU                     | sku_time                     |
|=========================|==============================|
```
11. Add simple product to Cart
12. Remove all products from the Cart
13. Copy masked_id from the quote_mask_id database table
14. When GraphQL API User prepares query variables:
 - optionQueryVariable = { "id": Int!, "value": [String!] }
in
```
{
	"input": {
		"cart_id": "$maskedQuoteId",
		"cartItems": [{
			"data": {
				"qty": 2,
				"sku": "simple"
			},
			"customizable_options": [<put option(s) here>]
		}]
	}
}
```
 for request message:
```
mutation addSimpleProductsToCart($input: AddSimpleProductsToCartInput) {
  addSimpleProductsToCart(input: $input) {
    cart {
      items {
        product {
          id
          name
          sku
        }
        qty
        ... on SimpleCartItem {
          customizable_options {
            id
            is_required
            type
            values {
              price {
                type
                units
                value
              }
              value
            }
          }
        }
      }
    }
  } 
}
```
and performs the request
### Expected result (*)

1. `{ "id": 57, "value": ["55"] }` is valid format for selecting customizable option of checkbox type and GraphQL API User succeeds with adding this option to cart

### Actual result (*)

1. For ` { "id": 57, "value": ["55"] }`
```
{
  "errors": [
    {
      "message": "Internal server error",
      "category": "internal",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "addSimpleProductsToCart"
      ]
    }
  ],
  "data": {
    "addSimpleProductsToCart": null
  }
}
```
2. For `{ "id": 57, "value": "55"}`
```
{
  "errors": [
    {
      "message": "The product's required option(s) weren't entered. Make sure the options are entered and try again.",
      "category": "graphql-input",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "addSimpleProductsToCart"
      ]
    }
  ],
  "data": {
    "addSimpleProductsToCart": null
  }
}
```

**Comment by TomashKhamlai** [comment]
Related to #413

**Comment by AlexandrKozyr** [comment]
reproduced locally.
problem is connected with:
1. the wrong scheme declaration 
```
input CustomizableOptionInput {
    id: Int!
    value: String!
}
```
2. but simply changing it to 
```
input CustomizableOptionInput {
    id: Int!
    value: [String!]!
}
```
makes impossible to add other values for customizable options.

So the problem is connected with a need to determine a properly customized option structure before adding it to the cart.
The first possible solution is to retrieve option details from M2. Something simialr to - https://github.com/magento/graphql-ce/pull/365/files 
The second option is to determine the option type with mutation and based on it make a proper options array.
```
input CustomizableOptionInput {
    id: Int!
    type: String!
    value: [String!]!
}
```

**Comment by eugene-shab** [comment]
Event: kharkivcd19

**Comment by eugene-shab** [comment]
As alternative we can use: 

```
input CustomizableOptionInput {
    id: Int!
    value: [String!]
}
```

After that in
`\Magento\QuoteGraphQl\Model\Cart\AddSimpleProductToCart::extractCustomizableOptions`
we can only check is it array or string and proceed as the default implementation.

**Comment by AlexandrKozyr** [comment]
@naydav 
we need your assistance.

**Comment by AlexandrKozyr** [comment]
we need to understand what format of the request for options 
```
input CustomizableOptionInput {
    id: Int!
    value: String!
}
```
then we need
customizable_options: [
              {id: 1, value: "2"},
              {id: 1, value: "1"}
            ]
this format to work and this issue is not relavant we need to solve the problem with in the https://github.com/magento/graphql-ce/pull/365
if not we need to change also the scheme to 
```
input CustomizableOptionInput {
    id: Int!
    value: [String!]!
}
```
and then solve the problem with the option that come as array.

**Comment by AlexandrKozyr** [comment]
@naydav 
current architecture docs
https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/add-items-to-cart/AddSimpleProductToCart.graphqls
makes this issue not relevant, because of the wrong query format. or we need to  change it

**Comment by naydav** [comment]
@TomashKhamlai Please verify ticket

**Comment by TomashKhamlai** [comment]
This issue is still valid because on the lowest level of abstraction we still expect an array but the string is returned. The current implementation do not allow to send `[{id: 1, value: "1"},{id: 1, value: "2"}]` neither `[{id: 1, value: "[1,2]"}]` or `[{id: 1, value: ["1", "2"]}]`

I don't mind to send values in different objects, like [{id: 1, value: "1"},{id: 1, value: "2"}], but the only problem is that in some queries we take into account only last value and here in #365 it is proposed to collect every value.

@AlexandrKozyr have you seen somewhere in requirements how multiple values with the same `id` should be treated? I would like to understand this idea and check some places where it is violated.

**Comment by AlexandrKozyr** [comment]
@TomashKhamlai 
Hi! as far as I got after the conversation with @naydav and @eugene-shab,  now there is the requirement in the architecture scheme that customized options values should be put as id-value pairs only, without sub-array
![image](https://user-images.githubusercontent.com/9674839/55232482-60f2aa00-522e-11e9-9e34-3a002fb04766.png)
https://github.com/magento/architecture/blob/master/design-documents/graph-ql/coverage/add-items-to-cart/AddSimpleProductToCart.graphqls
So, to be honest, we even almost change it to fix your bug but then we found another issue for this (https://github.com/magento/graphql-ce/issues/413) - where our colleague is trying to solve the issue another way (with one level array of options).
Answering your question: I don't have a kind of example of such a violation. But I believe we can check all places where the checkbox/multi-select functionality is used in Magento and determine does Architecture have the same solution for all such cases. Do this specific input type it clearly says that we can put only string in as the param.
I'll try to figure out this question.
